specs/CustomDownload.md

```diff
@@ -0,0 +1,771 @@
+                        UpdateProgress(download.get());
+                    }
+                    else
+                    {
+                        // If user selects `Cancel`, the download will be canceled.
+                        CHECK_FAILURE(args->put_Cancel(TRUE));
+                    }
+                };
+
+                // Obtain a deferral for the event so that the CoreWebView2
+                // doesn't examine the properties we set on the event args until
+                // after we call the Complete method asynchronously later.
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
+                CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                // This function can be called to show the download dialog and
+                // complete the event at a later time, allowing the developer to
+                // perform async work before the event completes.
+                m_completeDeferredDownloadEvent = [showDialog, deferral] {
+                    showDialog();
+                    CHECK_FAILURE(deferral->Complete());
+                };
+
+                return S_OK;
+            })
+        .Get(),
+    &m_downloadStartingToken));
+}
+
+// Update download progress
+void ScenarioCustomDownloadExperience::UpdateProgress(ICoreWebView2Download* download)
+{
+    // Register a handler for the `DownloadProgressSizeInBytesChanged` event.
+    CHECK_FAILURE(download->add_DownloadProgressSizeInBytesChanged(
+        Callback<ICoreWebView2DownloadProgressSizeInBytesChangedEventHandler>(
+            [this](ICoreWebView2Download* download, IUnknown* args) -> HRESULT {
+                // Here developer can update UI to show progress of a download using
+                // `download->get_DownloadProgressSizeInBytes` and
+                // `download->get_ExpectedDownloadSizeInBytes`.
+                return S_OK;
+            })
+        .Get(),
+    &m_downloadProgressSizeInBytesChangedToken));
+
+    // Register a handler for the `DownloadStateChanged` event.
+    CHECK_FAILURE(download->add_DownloadStateChanged(
+        Callback<ICoreWebView2DownloadStateChangedEventHandler>(
+            [this](ICoreWebView2Download* download,
+            ICoreWebView2DownloadStateChangedEventArgs* args) -> HRESULT {
+                COREWEBVIEW2_DOWNLOAD_STATE downloadState;
+                CHECK_FAILURE(download->get_State(&downloadState));
+                switch (downloadState)
+                {
+                case COREWEBVIEW2_DOWNLOAD_STATE_IN_PROGRESS:
+                    break;
+                case COREWEBVIEW2_DOWNLOAD_STATE_INTERRUPTED:
+                    // Here developer can take different actions based on `args->InterruptReason`.
+                    // For example, show an error message to the end user.
+                    CompleteDownload(download);
+                    break;
+                case COREWEBVIEW2_DOWNLOAD_STATE_COMPLETED:
```

Comment:
We can't complete until after the DownloadStarting event handler or deferral is completed because we won't know the destination file path which could result in errors. Document this along with event ordering.